### PR TITLE
Apply `black` to `astropy.io.ascii`

### PR DESCRIPTION
**Pull request by WilliamJamieson**
### Description

Apply `black` to `astropy.io.ascii` according to the process laid out by [APE 20](https://github.com/astropy/astropy-APEs/blob/main/APE20.rst).

### Checklist for package maintainer(s)

This checklist is meant to remind the package maintainer(s) who will review this pull request of some common things to look for. This list is not exhaustive.

- [x] Do the proposed changes actually accomplish desired goals?
- [x] Do the proposed changes follow the [Astropy coding guidelines](https://docs.astropy.org/en/latest/development/codeguide.html)?
- [x] Are tests added/updated as required? If so, do they follow the [Astropy testing guidelines](https://docs.astropy.org/en/latest/development/testguide.html)?
- [x] Are docs added/updated as required? If so, do they follow the [Astropy documentation guidelines](https://docs.astropy.org/en/latest/development/docguide.html#astropy-documentation-rules-and-guidelines)?
- [ ] Is rebase and/or squash necessary? If so, please provide the author with appropriate instructions. Also see ["When to rebase and squash commits"](https://docs.astropy.org/en/latest/development/when_to_rebase.html).
- [x] Did the CI pass? If no, are the failures related? If you need to run daily and weekly cron jobs as part of the PR, please apply the `Extra CI` label. Codestyle issues can be fixed by the [bot](https://docs.astropy.org/en/latest/development/workflow/development_workflow.html#pre-commit).
- [x] Is a change log needed? If yes, did the change log check pass? If no, add the `no-changelog-entry-needed` label. If this is a manual backport, use the `skip-changelog-checks` label unless special changelog handling is necessary.
- [x] Is this a big PR that makes a "What's new?" entry worthwhile and if so, is (1) a "what's new" entry included in this PR and (2) the "whatsnew-needed" label applied?
- [x] Is a milestone set? Milestone must be set but `astropy-bot` check might be missing; do not let the green checkmark fool you.
- [x] At the time of adding the milestone, if the milestone set requires a backport to release branch(es), apply the appropriate `backport-X.Y.x` label(s) *before* merge.

### CODE REVIEW

**Comment by dhomeier** [comment]
Thanks for setting this up! To start off with, a possibly very naïve question:
what is the difference between how black was applied in #13253 (8 files modified with a few score lines changed) and here (well, +3,517 −2,040...)?

A more general question mainly to @taldcroft and @hamogu:
do you think it helpful to add general style cleanups beyond those generated by `black` in this PR? As an example
```python
         raise ValueError(f'column(s) with dimension > {max_ndim} '
                          "cannot be be written with this format, try using 'ecsv' "
                          "(Enhanced CSV) format")
```
has been blackened to
```python
         raise ValueError(
             f'column(s) with dimension > {max_ndim} '
             "cannot be be written with this format, try using 'ecsv' "
             "(Enhanced CSV) format"
         )
```
without touching the single-quoted f-string (I thought it was actually the black default to put all strings in dquotes where possible, is this customised in our black configuration?). Anyway in this example the double quotes are required for the third line, so for the sake of consistent appearance they'd better be all put in double quotes.

I know there are good reasons to keep such manual fine-tuning separate from the automated formatting, but since we need to look through the entire changeset anyway, it would be a good occasion to do it here, and I at least may not have another close look at most of this code for quite a long time to come.

File: astropy/io/ascii/cds.py
Comment (dhomeier) [inline astropy/io/ascii/cds.py:135]: I know we are not meant to quarrel excessively with the way the formatting rules are imposed by black, because, they are the rules, but this, to me personally, is way beyond the border into the horrendous (a number more, similar examples in `core.py` below).
Do we already have some standards at what point we may or should apply `# fmt:` decorators according to https://github.com/astropy/astropy/pull/13949#discussion_r1010780782 to protect such constructs?

**Comment by WilliamJamieson** [comment]
> Thanks for setting this up! To start off with, a possibly very naïve question: what is the difference between how black was applied in #13253 (8 files modified with a few score lines changed) and here (well, +3,517 −2,040...)?

PR #13253 just added a black configuration but turned it off for all sub-packages. The file changes there were due to incompatibilities between our former `isort` configuration and `black`. This PR is the result of running black on `astropy.io.ascii`.
> 
> A more general question mainly to @taldcroft and @hamogu: do you think it helpful to add general style cleanups beyond those generated by `black` in this PR? As an example
> 
> ```python
>          raise ValueError(f'column(s) with dimension > {max_ndim} '
>                           "cannot be be written with this format, try using 'ecsv' "
>                           "(Enhanced CSV) format")
> ```
> 
> has been blackened to
> 
> ```python
>          raise ValueError(
>              f'column(s) with dimension > {max_ndim} '
>              "cannot be be written with this format, try using 'ecsv' "
>              "(Enhanced CSV) format"
>          )
> ```
> 
> without touching the single-quoted f-string (I thought it was actually the black default to put all strings in dquotes where possible, is this customised in our black configuration?). Anyway in this example the double quotes are required for the third line, so for the sake of consistent appearance they'd better be all put in double quotes.

Per APE 20 the initial `black` changes here are just those created by running `black --skip-string-normalization astropy/io/ascii`. This run does not normalize the string quotes in order to cut down on the initial diff to be reviewed (yes its a lot, but it would be even bigger if we normalized strings). Per APE 20, once we are happy with these changes I'll apply string normalization.

File: astropy/io/ascii/cds.py
Comment (WilliamJamieson) [reply astropy/io/ascii/cds.py:135]: Yes anything you feel is horrendous warrants a `# fmt:` decorator. Please feel free to push commit(s) adding `# fmt:` and undoing `black` formatting as you feel fit.

**Comment by WilliamJamieson** [comment]
Also note https://github.com/astropy/astropy/pull/13953#issuecomment-1300904451, in which we found running:

```shell
$ black --skip-string-normalization --preview astropy/io/ascii
```

may fix some of the less desirable string issues.

**Comment by dhomeier** [comment]
> Per APE 20 the initial `black` changes here are just those created by running `black --skip-string-normalization astropy/io/ascii`. This run does not normalize the string quotes in order to cut down on the initial diff to be reviewed (yes its a lot, but it would be even bigger if we normalized strings). Per APE 20, once we are happy with these changes I'll apply string normalization.

OK, good to know, then we don't need to bother with those types of extra tuning (I would have been happy to leave out the quote normalisation entirely otherwise, but I guess that cannot be taken out of APE 20).

**Comment by WilliamJamieson** [comment]
> OK, good to know, then we don't need to bother with those types of extra tuning (I would have been happy to leave out the quote normalisation entirely otherwise, but I guess that cannot be taken out of APE 20).

This is why pre-commit is failing right now. That bot can take care of normalization when this PR is ready for that.

File: astropy/io/ascii/cds.py
Comment (dhomeier) [inline astropy/io/ascii/cds.py:135]:
```suggestion
                col.start = int(re.sub(r'[-\s]', '',
                                       match.group('start') or match.group('end'))) - 1  # fmt: skip
```
OK, starting with a try if this is the correct syntax to preserve the original. I note that
```
                col.start = int(
                    re.sub(r'[-\s]', '', match.group('start') or match.group('end'))) - 1  # fmt: skip
```
would also stay within the new 88-char limit; not sure which is preferable in terms of readability.

Comment (taldcroft) [reply astropy/io/ascii/cds.py]: Reverting back here is fine, but I also stopped and looked at that line. IMO both versions are awkward but the Black version is better. In the original you have 3 nested parens and the end (always a bad sign) and a break after 2 of the 3 args to `re.sub()` which takes a lot of time to parse for me.  In the black version at least it is very easy to see that you have `int( some re.sub expression) - 1` (even if it does look weird).

Anyway, we knew this would happen, not a problem. :smile:

Comment (dhomeier) [reply astropy/io/ascii/cds.py]: I agree about the break in the middle of the args, though I hardly find it any easier to figure out where the `-1` belongs to in the black version with the 4th pair of `()` added!
Perhaps my alternative suggestion from https://github.com/astropy/astropy/pull/13956#discussion_r1012341124 would have been better, even if I'm not happy with the indentation level.

Yes, good to discuss some of these cases early on. 😁

**Review by hamogu — COMMENTED** [review]
I scrolled over the changes. In most cases, I'm OK with it either this way to that way, so I won't argue - the whole point of black was to just let black do it. I have not seen anything that's terrible after applying black. Even if I missed something ans someone finds a piece of code hard ot read when working on it later,  formatting can always be changed later.

File: astropy/io/ascii/tests/test_cds.py
Comment (WilliamJamieson) [inline astropy/io/ascii/tests/test_cds.py:480]: Looks like something went wrong with this table. I think the spaces at the end of the line were removed incorrectly.

File: astropy/io/ascii/tests/test_cds.py
Comment (WilliamJamieson) [inline astropy/io/ascii/tests/test_cds.py:528]: I think the same thing maybe happening here

**Review by WilliamJamieson — COMMENTED** [review]
I noticed the most recent changes are causing the unit tests to fail. I think the issue is the removal of trailing spaces in some of the test strings.

File: astropy/io/ascii/tests/test_cds.py
Comment (taldcroft) [reply astropy/io/ascii/tests/test_cds.py:480]: I'll fix that. The tests should be written in a way to not depend on trailing spaces. I've done that in other contexts.

**Review by taldcroft — APPROVED** [review]
From my perspective this is good to go. I suppose @dhomeier or @hamogu should look through the additional commits I made, which are much easier to review than the original changes.

**Comment by hamogu** [comment]
As I said before, I don't have strong opinions either way; the improvements made by @taldcroft make the code more readable and remove some stupid cases. However, even if we miss a few of those cases, no harm done because the AST is not changed, the tests pass (still running now, but hopefully will) and, if we ever look at a specific piece of code again and it looks oddly formatted, we can always change it later without risking to break backwards-compatibility. In that sense, this is probably the lowest risk PR I've lever look at.

I have no further comments, but won't approve until tests pass (that also gives @dhomeier some more time to look).

**Comment by taldcroft** [comment]
@hamogu - I agree fully with your perspective, and tests are passing now!

**Comment by dhomeier** [comment]
From the (few) files I've reviewed in detail so far, things are looking much improved with @taldcroft's updates. I'll go through the rest today.

**Review by dhomeier — CHANGES_REQUESTED** [review]
Added some more suggestions mainly for better usage of f-strings.
With those reviewed and tests passing I am ready to give this a go, too; the rest I only dislike about as much as anticipated.
`        ):`

File: astropy/io/ascii/cds.py
Comment (dhomeier) [inline astropy/io/ascii/cds.py:130]:
```suggestion
                col.start = int(
                    re.sub(r'[-\s]', '', match.group('start') or match.group('end'))) - 1  # fmt: skip
```
alternative following discussion in https://github.com/astropy/astropy/pull/13956#discussion_r1012412088

File: astropy/io/ascii/cds.py
Comment (dhomeier) [inline astropy/io/ascii/cds.py:131]:
```suggestion
```

**Review by dhomeier — CHANGES_REQUESTED** [review]

File: astropy/io/ascii/cds.py
Comment (taldcroft) [reply astropy/io/ascii/cds.py:44]: @dhomeier - one of the goals of the transitions was no code changes, which gives us confidence in not breaking anything. 

There are automated tools we can use later to make changes like this to replace `.format()` method calls where it makes sense. So I would prefer to NOT make any f-string changes as part of the black transition. You can certainly do so as a post-black update if you like.

File: astropy/io/ascii/core.py
Comment (taldcroft) [reply astropy/io/ascii/core.py:334]: Same comment, no code changes please.

**Comment by dhomeier** [comment]
I think I misinterpreted the effect of some of your changes following my initial question, especially the merging of strings and stripping whitespace in 1ab9400. Should probably strip this down then on the moved comments and a few others.
Re. the automated f-string conversion, I thought that had already been applied some time ago, so the remaining `.format()` instances would somehow have been avoided by the tools? Can still probably do a general grep or something through the remaining code.

**Review by dhomeier — APPROVED** [review]

**Comment by dhomeier** [comment]
I've commited the non code-changing part only and suggest to ignore the rest (in lieu of a batch-mark-resolved).

File: astropy/io/ascii/tests/test_cds.py
Comment (dhomeier) [inline astropy/io/ascii/tests/test_cds.py:550]: @taldcroft I generally agree with your emailed suggestion to accept trivial code changes in these PRs, especially where already committed. This also seemed an inconspicuous change merely getting the test to work again.
But looking at this again today I am having second thoughts – if I recall correctly, the current behaviour of the MRT writer is to pad all header lines to the width of the longest content (non-blank) string. This does not actually seem to be mandated by the MRT standard, so in principle the writer would be free to just put out the stripped lines.
But the current `astropy.io.ascii` functionality is to create a header padded to constant line widths (I think implemented to make parsing more convenient), and by now people may have come to rely on that property. The previous test was implicitly also testing this, and I'd feel better if we kept it that way and protect the blank space in `exp_output` from black instead, as @WilliamJamieson initially proposed.

Comment (dhomeier) [reply astropy/io/ascii/tests/test_cds.py:550]: Ah, I thought that last point [had been discussed](https://github.com/astropy/astropy/pull/12302#issuecomment-955086901) before!

Comment (taldcroft) [reply astropy/io/ascii/tests/test_cds.py]: Fixed

**Comment by taldcroft** [comment]
@dhomeier - I reviewed all of your code-changing suggestions and they are all correct (including the QDP fix for the missing comma), so following the new process I committed them here. You should review my new commits as a cross-checkl

**Review by taldcroft — APPROVED** [review]

File: astropy/io/ascii/tests/test_c_reader.py
Comment (dhomeier) [inline astropy/io/ascii/tests/test_c_reader.py:326]:
```suggestion
  a, b,   c
  
"""
```
Got the newline in the wrong place here. Obviously it makes no difference to the test, but since the original produced a whitespace line at the end, it's probably best to reproduce that.

**Comment by dhomeier** [comment]
I agree with the updates to the CDS tests and the procedure; just one fix to my own fix to `test_lstrip_whitespace` left.

File: astropy/io/ascii/tests/test_cds_header_from_readme.py
Comment (taldcroft) [reply astropy/io/ascii/tests/test_cds_header_from_readme.py:149]: Highly unlikely...

File: astropy/io/ascii/tests/test_c_reader.py
Comment (taldcroft) [inline astropy/io/ascii/tests/test_c_reader.py:325]: @dhomeier - it looks like this change has introduced trailing whitespace and caused pre-commit to file. Can you please just remove this commit?

Comment (dhomeier) [reply astropy/io/ascii/tests/test_c_reader.py:325]: Adding trailing whitespace was probably the point of the original `str + '  \n'` construct.
Any ideas how to implement this in a way that will not let pre-commit fail?

Comment (taldcroft) [reply astropy/io/ascii/tests/test_c_reader.py:325]: I presume that the code that black generated prior to your fix in 9f0450e was fine?
```
    text = """
     1,  2,   \t3
 A,\t\t B,  C
  a, b,   c  \n
"""
```

Comment (dhomeier) [reply astropy/io/ascii/tests/test_c_reader.py:325]: It was passing, but testing a different string (ending in `' a, b,   c  \n\n'`, whereas originally it was `' a, b,   c\n  \n'`).
Feel free to delete the commit, I can't find an option to do it online and will be away from my computer for a few hours. I'd just try the alternative option afterwards, as I don't know what exactly is happening in the pre-commit.

**Comment by taldcroft** [comment]
Passing all tests, good to go!!

**Pull request merged**
